pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -0,0 +1,337 @@
+package sharediterator
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.uber.org/zap"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/storage/storagewrappers/storagewrappersutil"
+)
+
+var (
+	tracer = otel.Tracer("openfga/pkg/storagewrappers/sharediterator")
+
+	_ storage.RelationshipTupleReader = (*IteratorDatastore)(nil)
+	_ storage.TupleIterator           = (*sharedIterator)(nil)
+
+	sharedIteratorQueryHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "shared_iterator_query_ms",
+		Help:                            "The duration (in ms) of a shared iterator query labeled by operation and shared.",
+		Buckets:                         []float64{1, 2, 5, 10, 25, 50, 80, 100, 150, 200, 300, 1000},
```

Comment:
Remember, buckets == tag cardinality. 
We can survive with less granularity and still see the value.